Update @swift3_migration attributes in swift-2.3-branch

benlangmuir opened the pull request:
<!-- Please complete this template before creating the pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->
This updates the migration attributes to match the latest Swift 3.0
stdlib and SDK overlays. It has no affect on compilation, only on
migration.

rdar://problem/27751175

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

A smoke test on macOS does the following:

1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:

1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

benlangmuir commented:
@swift-ci Please test

benlangmuir commented:
> 14:09:12 /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/hudson2212060325189275032.sh: line 21: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-2.3-branch/swift/utils/update-checkout: No such file or directory

I guess 2.3 pull-request testing isn't working?  cc @lplarson  @shahmishal

tkremenek commented:
@swift-ci test os x

tkremenek commented:
@benlangmuir This is my fault.  Fixing.

gribozavr commented on stdlib/public/core/Arrays.swift.gyb:
```diff
@@ -508,7 +508,7 @@ extension ${Self} : _ArrayType {
   /// Construct a ${Self} of `count` elements, each initialized to
   /// `repeatedValue`.
-  @swift3_migration(renamed="init(repeating:count:)")
+  @swift3_migration(message="Renamed to init(repeating:count:) and reordered parameters")
```
I thought `renamed=""` could handle parameter reordering.  If it can't, ignore this comment.

benlangmuir replied:
Not in swift 2 at least

jrose-apple replied:
It definitely can't; how would it know which parameters got reordered? Or would it assume any labels that didn't change are reordered rather than renamed?

gribozavr commented on stdlib/public/core/CString.swift:
```diff
@@ -36,6 +37,7 @@ extension String {
   /// Returns `nil` if the `CString` is `NULL`.  If `CString` contains
   /// ill-formed UTF-8 code unit sequences, replaces them with replacement
   /// characters (U+FFFD).
+  @swift3_migration(message="Please use String.init(cString:). Note however that is does not accept NULL as a valid pointer. String.decodeCString might also be an option, should you need more control.")
```
s/NULL/nil/ ?

gribozavr commented on stdlib/public/core/CollectionAlgorithms.swift.gyb:
```diff
@@ -141,6 +144,7 @@ extension MutableCollectionType
   ///
 ${orderingRequirementForComparable}
+  @swift3_migration(message="Slice the collection using the range, and call partition()")
   public mutating func partition(range: Range<Index>) -> Index {
```
Actually, there is no direct replacement for this API.  The argument-less partition was removed.  You can emulate the Swift 2 partition() in Swift 3 by providing a closure that compares the parameter against an arbitrary element of the collection (e.g., the first element).

benlangmuir replied:
Do you have a suggestion for the message, or should we just remove this attribute?

gribozavr replied:
I think pointing developers at `partition(by:)` (and its documentation?) would be good enough.

benlangmuir replied:
How about: Slice the collection using the range, and call partition(by:) with an appropriate closure.

gribozavr replied:
Sounds good.

gribozavr commented on stdlib/public/core/CollectionAlgorithms.swift.gyb:
```diff
@@ -127,6 +129,7 @@ extension MutableCollectionType where Index : RandomAccessIndexType {
 ${partitionDocComment}
   ///
 ${orderingRequirementForPredicate}
+  @swift3_migration(message="Slice the collection using the range, and call partition(isOrderedBefore:)")
```
`partition(by:)`

benlangmuir replied:
Nice catch.

gribozavr commented on stdlib/public/core/Index.swift:
```diff
@@ -92,7 +92,7 @@ public postfix func ++ <T : _Incrementable> (inout i: T) -> T {
 /// successor, if any, is reachable by applying the value's
 /// `successor()` method.
-@swift3_migration(renamed="ForwardIndex")
+@swift3_migration(message="ForwardIndex methods are now available from the Collection protocol.")
```
Could you also add something like "In Swift 3, collection indices don't need to implement ForwardIndex, but only required to implement Comparable." to push the user toward the right migration.

gribozavr replied:
And same for `BidirectionalIndex` and `RandomAccessIndex`.

benlangmuir replied:
Is the requirement the same in all cases - just implement Comparable?

gribozavr replied:
Right.

gribozavr commented on stdlib/public/core/Repeat.swift:
```diff
@@ -25,7 +25,7 @@ public struct Repeat<Element> : CollectionType {
   /// value `repeatedValue`.
-  @swift3_migration(renamed="repeatElement")
+  @swift3_migration(renamed="repeatElement(_:count:)")
   public init(count: Int, repeatedValue: Element) {
```
Looks like this API needs parameter reordering, should it use `message=` instead?

gribozavr commented on stdlib/public/core/StaticString.swift:
```diff
@@ -131,7 +131,7 @@ public struct StaticString
 
   /// Returns a `String` representing the same sequence of Unicode
   /// scalar values as `self` does.
-  @swift3_migration(message="use the 'String(_:)' initializer")
```
Since `StaticString` does not conform to `LosslessLiteralConvertible`, you need to call `String(describing:)`.

benlangmuir replied:
Huh, I'm not sure why this attribute was removed at all, since stringValue was really removed.  So `message="Use the String(describing:) initializer"` would probably be appropriate here.

benlangmuir replied:
Oh, I didn't see this was a change, not a removal.  Okay.

gribozavr commented on stdlib/public/core/UnicodeScalar.swift:
```diff
@@ -34,6 +34,7 @@ public struct UnicodeScalar :
   }
 
   /// Creates an instance of the NUL scalar value.
+  @swift3_migration(message="use 'UnicodeScalar(\"\\0\")'")
```
`UnicodeScalar("\0")!`, but `"\0" as UnicodeScalar` would be better since it does not require force-unwraps.

gribozavr commented:
LGTM modulo comments, thank you!

benlangmuir commented:
@swift-ci Please test

benlangmuir commented:
@swift-ci Please test

benlangmuir commented:
This failure looks unrelated to my change.
> /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-2.3-branch/swift/stdlib/public/SDK/Contacts/Contacts.swift:19:21: error: protocol '__BridgedNSError' requires '_NSErrorDomain' to be available on OS X 10.9.0 and newer
16:57:06   public static var _NSErrorDomain: String { return CNErrorDomain }
16:57:06                     ^

And the whole extension is marked with availability, so I would have expected this to be okay....
```
@available(OSX, introduced=10.11)
@available(iOS, introduced=9.0)
extension CNErrorCode : _BridgedNSError {
  public static var _NSErrorDomain: String { return CNErrorDomain }
}
```

benlangmuir commented:
Linux failure is a configuration issue:
> 17:48:23 Error: Unknown setting: jobs
> 17:48:23 Usage: build-script-impl [--help|-h] [ --SETTING=VALUE | --SETTING VALUE | --SETTING ]*

benlangmuir commented:
I was able to reproduce the error from the macOS tests locally and confirm it was *not* caused by my patch.  It seems this preset `buildbot_all_platforms,tools=RA,stdlib=RA` is failing, but the incremental one we use in regular CI is passing.  I haven't figured out why yet.

benlangmuir commented:
(FYI Devin figured out the build failure: the PR bot is building with swift-2.3-branch of clang/llvm, but (confusingly) should be using swift-3.0-branch).  Mishal is working on a fix.
